Add hooks structure for passing around user provided hooks, add a new guard_failure_fn

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90370
* #90130
* __->__ #90129

This PR introduces a new function we can pass to `torch._dynamo.optimize` - guard_failure_fn. Usage is in the PR, and the one stacked on top of it, but the gist of it is that it emits failed guard `reason` strings alongside code. This is useful for tests and debugging, as it gives far finer grained assertions and control than the compile counter alone. 

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
Guard debug infra

Test it


undo

rm bp

[ghstack-poisoned]

ezyang commented on test/dynamo/test_misc.py:
```diff
@@ -2973,6 +2973,35 @@ def fn(x, y):
+
+        self.assertTrue(guard_failure is not None)
+        self.assertEqual(guard_failure[0][0], "k == 3")
+        self.assertTrue("dynamo/test_misc.py" in guard_failure[1].co_filename)
```
seems unlikely to work on Windows

voznesenskym replied:
Agreed. Will rm.

ezyang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -361,6 +363,7 @@ def _compile(
     one_graph: bool,
     export: bool,
     guard_export_fn: Optional[Callable[[Set[Guard]], None]] = None,
+    guard_fail_fn: Optional[Callable[[Tuple[str, str]], None]] = None,
```
Consider defining a NamedTuple here to make it more clear what the various strs mean; the type annotation here is useless.

voznesenskym replied:
Sure

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -869,6 +875,13 @@ def guard_fail_hook(
+    except Exception as e:
+        log.error(
+            "Failure in guard_fail_fn callback - raising here will cause a NULL Error on guard eval"
+        )
```
exc_info=True

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -854,7 +860,7 @@ def guard_fail_hook(
     """
-    if not last:
+    if not guard_fn.guard_fail_fn and not last:
         return
```
This condition doesn't seem right.

Ordinarily, we only update stats when last is true. However, if I set `guard_fail_fn`, I will update stats for EVERY guard failure, even if it is not last. You probably don't want this behavior change?

voznesenskym replied:
Oh, I do, but would be open to changing it. I figured having the stats reflect how many calls the user saw to this would keep things a little more uniform. No strong opinion tho.

ezyang replied:
So first, if you are intentionally making a behavior change, call it out!

I don't actually know what the most desirable stats for guard failures is. It does seem weird, though, if you have something like five distinct guarded code objects, and at least one of them hits on all your runs, you will still keep accumulating guard failure stats for all the code objects you tested but failed a guard on (but it passed later).

voznesenskym replied:
Yes - but ideally, if you passed in a guard_fn.guard_fail_fn, it would be more jarring if these were not congruent, right? I dont know which is less jarring. Let's just make a call and add a comment.

wconstab replied:
yea this is tricky.

the "last" business refers to us testing and failing the final code object we have in our cache.  If there are many functions in our cache, it is not likely that we care particularly more about the reasons we fail the last one vs one of the previous ones, depending what we're debugging.

it was definitely too verbose to print the failures for every code object which might have driven me to do this "last" thing.  But if we are now able to return a dict i think it makes sense for it to contain all the failure info.

A more useful 'print' version of this might perform a sort over all the code objects, ranking by which one failed the least guards, in hopes that it finds an 'almost useful' function and guides the user to which part of their model code caused that to not be useful.  And that could be built on top of the dict from the guard_fail_fn iiuc, which is cool.

voznesenskym commented:
> Having to plumb all the debug functions around the code everywhere kwarg-by-kwarg seems suboptimal. Maybe we should put them in a hooks data structure or something?

Sure, no concern with that.

jansel commented on torch/_dynamo/convert_frame.py:
```diff
@@ -267,6 +267,7 @@ def exception_handler(e, code, frame=None):
 def convert_frame_assert(
     compiler_fn: CompilerFn,
     guard_export_fn=None,
+    guard_fail_fn=None,
```
Let's introduce some sort of options/hooks data structure, so we don't need so much plumbing to pass these around.

voznesenskym replied:
Done

voznesenskym committed:
Update on "Guard debug infra"


This PR introduces a new function we can pass to `torch._dynamo.optimize` - guard_failure_fn. Usage is in the PR, and the one stacked on top of it, but the gist of it is that it emits failed guard `reason` strings alongside code. This is useful for tests and debugging, as it gives far finer grained assertions and control than the compile counter alone. 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/guards.py:
```diff
@@ -848,28 +865,44 @@ def id_ref(self, obj):
     scope = {rename_implicit(k): v for k, v in f_locals.items()}
     scope.update(guard_fn.closure_vars)
-    reasons = []
+    reason = None
```
@wconstab why was this ever a list?

voznesenskym committed:
Update on "Add hooks structure for passing around user provided hooks, add a new guard_failure_fn"


This PR introduces a new function we can pass to `torch._dynamo.optimize` - guard_failure_fn. Usage is in the PR, and the one stacked on top of it, but the gist of it is that it emits failed guard `reason` strings alongside code. This is useful for tests and debugging, as it gives far finer grained assertions and control than the compile counter alone. 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/hooks.py:
```diff
@@ -0,0 +1,9 @@
+@dataclasses.dataclass
+class Hooks:
+    guard_export_fn: Optional[Callable[[Set["Guard"]], None]]
+    guard_fail_fn: Optional[Callable[[Tuple["GuardFail"]], None]]
```
Circular ref :(

ezyang replied:
I'd recommend defining Hooks in guards.py, that makes more sense anyway

voznesenskym replied:
Yeah, that's fine

voznesenskym commented on torch/_dynamo/convert_frame.py:
```diff
@@ -458,9 +465,9 @@ def transform(instructions, code_options):
 
 
-def convert_frame(compiler_fn: CompilerFn, guard_export_fn=None):
+def convert_frame(compiler_fn: CompilerFn, hooks=None):
```
It's Hooks, its not optional after a few entry points, and we pass it around with an empty default to be sure.

ezyang replied:
it's Optional[Hooks] here...

voznesenskym replied:
Yes, this is wrong :)

voznesenskym committed:
Update on "Add hooks structure for passing around user provided hooks, add a new guard_failure_fn"


This PR introduces a new function we can pass to `torch._dynamo.optimize` - guard_failure_fn. Usage is in the PR, and the one stacked on top of it, but the gist of it is that it emits failed guard `reason` strings alongside code. This is useful for tests and debugging, as it gives far finer grained assertions and control than the compile counter alone. 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/hooks.py:
```diff
@@ -0,0 +1,9 @@
+
+
+@dataclasses.dataclass
+class Hooks:
```
It's not just for guards, and its already in _dynamo...

voznesenskym committed:
Update on "Add hooks structure for passing around user provided hooks, add a new guard_failure_fn"


This PR introduces a new function we can pass to `torch._dynamo.optimize` - guard_failure_fn. Usage is in the PR, and the one stacked on top of it, but the gist of it is that it emits failed guard `reason` strings alongside code. This is useful for tests and debugging, as it gives far finer grained assertions and control than the compile counter alone. 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
